lib/rs/src/impls.rs

```diff
@@ -0,0 +1,195 @@
+pub use protocol::{self, Encode, Decode, Type, ThriftTyped};
+pub use {Protocol, Transport, Result, Error};
+
+pub use std::collections::{HashSet, HashMap};
+pub use std::hash::Hash;
+
+impl ThriftTyped for bool { fn typ() -> Type { Type::Bool } }
+impl ThriftTyped for i8  { fn typ() -> Type { Type::Byte } }
+impl ThriftTyped for i16 { fn typ() -> Type { Type::I16 } }
+impl ThriftTyped for i32 { fn typ() -> Type { Type::I32 } }
+impl ThriftTyped for i64 { fn typ() -> Type { Type::I64 } }
+impl ThriftTyped for f64 { fn typ() -> Type { Type::Double } }
+impl ThriftTyped for () { fn typ() -> Type { Type::Void } }
+impl ThriftTyped for String { fn typ() -> Type { Type::String } }
+impl<T: ThriftTyped> ThriftTyped for Vec<T> { fn typ() -> Type { Type::List } }
+impl<T: ThriftTyped> ThriftTyped for Option<T> { fn typ() -> Type { T::typ() } }
+impl<T: ThriftTyped> ThriftTyped for HashSet<T> { fn typ() -> Type { Type::Set } }
+impl<K: ThriftTyped, V: ThriftTyped> ThriftTyped for HashMap<K, V> { fn typ() -> Type { Type::Map } }
+
+impl<X: Encode> Encode for Vec<X> {
```

Comment:
I really do not think it is appropriate to use Encode and Decode for these operations, as these names do not imply any kind of transmission - they are used to describe a change in the encoding. So they would be ok in a protocol, but these actions combine encoding and sending.

If we cannot use Write and Read maybe they should be called ThriftWrite and ThriftRead.